lib/Interpreter/TransactionPool.h

```diff
@@ -20,74 +20,59 @@ namespace clang {
 
 namespace cling {
   class TransactionPool {
-#define TRANSACTIONS_IN_BLOCK 8
-#define POOL_SIZE 2 * TRANSACTIONS_IN_BLOCK
-  private:
+    enum {
+      kTransactionsInBlock = 8,
+      kPoolSize = 2 * kTransactionsInBlock
+    };
+
     // It is twice the size of the block because there might be easily around 8
     // transactions in flight which can be empty, which might lead to refill of
     // the smallvector and then the return for reuse will exceed the capacity
     // of the smallvector causing redundant copy of the elements.
     //
-    llvm::SmallVector<Transaction*, POOL_SIZE>  m_Transactions;
-
-    ///\brief The Sema required by cling::Transactions' ctor.
-    ///
-    clang::Sema& m_Sema;
+    llvm::SmallVector<Transaction*, kPoolSize>  m_Transactions;
 
-    // We need to free them in blocks.
-    //
-    //llvm::SmallVector<Transaction*, 64> m_TransactionBlocks;
-#ifndef NDEBUG
-    bool m_Debug;
-#endif
-
-  private:
   public:
-    TransactionPool(clang::Sema& S) : m_Sema(S) {
-#ifndef NDEBUG
-      m_Debug = false;
-#endif
-    }
-
+    TransactionPool() {}
     ~TransactionPool() {
+      // Only free the memory as anything put in m_Transactions will have
+      // already been destructed in releaseTransaction
       for (size_t i = 0, e = m_Transactions.size(); i < e; ++i)
-        delete m_Transactions[i];
+        ::operator delete(m_Transactions[i]);
     }
 
-    Transaction* takeTransaction() {
-      if (m_Transactions.empty())
-        return new Transaction(m_Sema);
-      Transaction* T = new (m_Transactions.pop_back_val()) Transaction(m_Sema);
-#ifndef NDEBUG
-      // *Very useful for debugging purposes and setting breakpoints in gdb.
-      if (m_Debug)
-        T = new Transaction(m_Sema);
-#endif
+    Transaction* takeTransaction(clang::Sema& S) {
+      Transaction *T;
+      if (m_Transactions.empty()) {
+        T = (Transaction*) ::operator new(sizeof(Transaction));
+        new(T) Transaction(S);
+      } else
+        T = new (m_Transactions.pop_back_val()) Transaction(S);
 
-      T->m_State = Transaction::kCollecting;
       return T;
     }
 
-    void releaseTransaction(Transaction* T) {
-      assert((T->getState() == Transaction::kCompleted ||
-              T->getState() == Transaction::kRolledBack)
-             && "Transaction must completed!");
+    void releaseTransaction(Transaction* T, bool reuse = true) {
```

Comment:
We cat set watch points if transaction number N always ends up at a given address. If they are re-used then the transaction N is at the same memory address as a gazillion transactions before. That makes watch points much more difficult to use...